src/analyses/goto_check.cpp

```diff
@@ -1853,27 +1851,44 @@ optionalt<exprt> goto_checkt::rw_ok_check(exprt expr)
 
 /// Set a Boolean flag to a new value (via `set_flag`) and restore the previous
 /// value when the entire object goes out of scope.
+/// Tracks flags to allow detecting double sets
 class flag_resett
 {
 public:
-  /// Store the current value of \p flag and then set its value to \p new_value.
-  void set_flag(bool &flag, bool new_value)
-  {
+  /// \brief Store the current value of \p flag and
+  /// then set its value to \p new_value.
+  /// Triggers an INVARIANT iff the flag is set
+  /// more than once with different values.
+  void
+  set_flag(bool &flag, bool new_value, std::function<std::string(void)> err_msg)
+  {
+    INVARIANT(!(seen(flag) && flag != new_value), err_msg());
+    seen_flags.insert(&flag);
     if(flag != new_value)
     {
       flags_to_reset.emplace_back(&flag, flag);
       flag = new_value;
     }
+    return;
   }
 
-  /// Restore the values of all flags that have been modified via `set_flag`.
+  /// \brief True iff flag has already been seen by \ref set_flag
+  bool seen(bool &flag)
+  {
+    return seen_flags.find(&flag) != seen_flags.end();
```

Comment:
Hmm, we seem to be taking the address of `flag` here. So if we have two `flag`s with the same content but different addresses then this would break?

Do we have some mechanism to guard against it? Or if there's some other reason as to why this is a non-issue, we should mention that in the comments above this function.

Reply:
the solution would be to make `flag_resett` a private class inside `goto_checkt` to ensure it is not going to be misused elsewhere.